File: track.lua
-- Track geometry, surface zones, curbs, and trees (pure logic, no Love2D dependency)
-- Now supports spline-based tracks from configuration

local track = {}
local sqrt, floor, atan2, max = math.sqrt, math.floor, math.atan2, math.max
local huge, pi = math.huge, math.pi

-- Local seeded PRNG to avoid corrupting global math.random state
local function makeRng(seed)
    local s = seed % 2147483647
    if s <= 0 then s = s + 2147483646 end
    local rng = {}
    function rng:next()
        s = (s * 48271) % 2147483647
        return s / 2147483647
    end
    function rng:range(lo, hi)
        return lo + self:next() * (hi - lo)
    end
    function rng:int(lo, hi)
        return floor(self:range(lo, hi + 0.999))
    end
    return rng
end

local spline = require("spline")
local catmullRom = spline.catmullRom

-- Generate smooth path from control points using Catmull-Rom spline
local function generateSplinePath(controlPoints, segmentsPerCurve)
    segmentsPerCurve = segmentsPerCurve or 20
    local path = {}
    local n = #controlPoints
    
    for i = 1, n do
        local p0 = controlPoints[((i - 2) % n) + 1]
        local p1 = controlPoints[((i - 1) % n) + 1]
        local p2 = controlPoints[(i % n) + 1]
        local p3 = controlPoints[((i + 1) % n) + 1]
        
        for j = 0, segmentsPerCurve - 1 do
            local t = j / segmentsPerCurve
            local point = catmullRom(p0, p1, p2, p3, t)
            table.insert(path, point)
        end
    end
    
    return path
end

-- Calculate path length and cumulative distances
local function calculatePathMetrics(path)
    local totalLength = 0
    local cumulative = {0}
    
    for i = 2, #path do
        local dx = path[i].x - path[i-1].x
        local dy = path[i].y - path[i-1].y
        totalLength = totalLength + sqrt(dx * dx + dy * dy)
        cumulative[i] = totalLength
    end
    
    -- Close the loop
    local dx = path[1].x - path[#path].x
    local dy = path[1].y - path[#path].y
    totalLength = totalLength + sqrt(dx * dx + dy * dy)
    
    return totalLength, cumulative
end

-- Get point at a specific percentage along the path
local function getPointAtPercent(path, cumulative, totalLength, pct)
    local targetDist = pct * totalLength
    
    for i = 2, #path do
        if cumulative[i] >= targetDist then
            local prevDist = cumulative[i-1]
            local segmentLength = cumulative[i] - prevDist
            local t = (targetDist - prevDist) / segmentLength
            
            return {
                x = path[i-1].x + t * (path[i].x - path[i-1].x),
                y = path[i-1].y + t * (path[i].y - path[i-1].y)
            }
        end
    end
    
    return path[#path]
end

-- Get tangent direction at a path index
local function getTangent(path, index)
    local n = #path
    local prev = path[((index - 2) % n) + 1]
    local nextPt = path[(index % n) + 1]

    local dx = nextPt.x - prev.x
    local dy = nextPt.y - prev.y
    local len = sqrt(dx * dx + dy * dy)
    
    if len > 0 then
        return dx / len, dy / len
    end
    return 1, 0
end

local function generateCurbs()
    track.outerCurbs = {}
    track.innerCurbs = {}

    local step = max(1, floor(#track.centerPath / 80))
    local index = 0

    for i = 1, #track.centerPath, step do
        local outer = track.outerPath[i]
        local inner = track.innerPath[i]
        local tx, ty = getTangent(track.centerPath, i)
        local angle = atan2(ty, tx)

        table.insert(track.outerCurbs, { x = outer.x, y = outer.y, angle = angle, index = index })
        table.insert(track.innerCurbs, { x = inner.x, y = inner.y, angle = angle, index = index })
        index = index + 1
    end
end

local function generateTrees()
    track.trees = {}
    local rng = makeRng(77)

    -- Generate trees around the track
    for _ = 1, 25 do
        local attempts = 0
        while attempts < 50 do
            local x = rng:int(20, 780)
            local y = rng:int(20, 580)

            -- Check if far enough from track
            local minDist = huge
            for _, p in ipairs(track.centerPath) do
                local dx = x - p.x
                local dy = y - p.y
                local dist = sqrt(dx * dx + dy * dy)
                if dist < minDist then
                    minDist = dist
                end
            end

            -- Place tree if it's off the track
            if minDist > track.width / 2 + 15 then
                table.insert(track.trees, {
                    x = x, y = y,
                    trunkH = 6 + rng:next() * 4,
                    canopyR = 8 + rng:next() * 7,
                    green = 0.3 + rng:next() * 0.3,
                    shade = 0.1 + rng:next() * 0.1
                })
                break
            end
            attempts = attempts + 1
        end
    end
end

local function generateSurfaceZones(zoneConfigs)
    track.surfaceZones = {}

    if zoneConfigs then
        for _, z in ipairs(zoneConfigs) do
            table.insert(track.surfaceZones, {
                startPct = z.startPct,
                endPct = z.endPct,
                grip = z.grip,
                bumpiness = z.bumpiness,
                name = z.name,
                color = z.color or {0.5, 0.5, 0.5, 0.0}
            })
        end
    else
        -- Default zones
        table.insert(track.surfaceZones, {
            startPct = 0.0, endPct = 1.0,
            grip = 0.95, bumpiness = 0.05,
            name = "Smooth Tarmac",
            color = {0.5, 0.5, 0.5, 0.0}
        })
    end
end

-- Initialize track from a configuration
function track.initFromConfig(config)
    track.config = config
    track.name = config.name
    track.width = config.width
    
    -- Generate the center line path
    track.centerPath = generateSplinePath(config.points, 25)
    track.pathLength, track.cumulative = calculatePathMetrics(track.centerPath)
    
    -- Generate inner and outer boundaries
    track.innerPath = {}
    track.outerPath = {}
    local halfWidth = config.width / 2
    
    for i, p in ipairs(track.centerPath) do
        local tx, ty = getTangent(track.centerPath, i)
        -- Normal is perpendicular to tangent
        local nx, ny = -ty, tx
        
        table.insert(track.innerPath, {
            x = p.x + nx * halfWidth,
            y = p.y + ny * halfWidth
        })
        table.insert(track.outerPath, {
            x = p.x - nx * halfWidth,
            y = p.y - ny * halfWidth
        })
    end
    
    -- Set up finish line at the start of the track
    local startPoint = track.centerPath[1]
    local tx, ty = getTangent(track.centerPath, 1)
    local nx, ny = -ty, tx
    
    track.finishX = startPoint.x
    track.finishY1 = startPoint.y + ny * halfWidth
    track.finishY2 = startPoint.y - ny * halfWidth
    track.finishAngle = atan2(ty, tx)

    -- Finish line as proper 2D geometry for direction-independent crossing detection
    track.finishPoint = { x = startPoint.x, y = startPoint.y }
    track.finishForward = { x = tx, y = ty }  -- track tangent = forward direction
    track.finishP1 = { x = startPoint.x + nx * halfWidth, y = startPoint.y + ny * halfWidth }
    track.finishP2 = { x = startPoint.x - nx * halfWidth, y = startPoint.y - ny * halfWidth }
    
    -- For compatibility: approximate center for trees
    local sumX, sumY = 0, 0
    for _, p in ipairs(config.points) do
        sumX = sumX + p.x
        sumY = sumY + p.y
    end
    track.cx = sumX / #config.points
    track.cy = sumY / #config.points
    
    -- Generate curbs, trees, and surface zones
    generateCurbs()
    generateTrees()
    generateSurfaceZones(config.surfaceZones)
    
    -- Store start position and angle for car init
    track.startX = startPoint.x
    track.startY = startPoint.y
    track.startAngle = config.startAngle or track.finishAngle
end

-- Legacy init for backward compatibility (creates oval track)
function track.init()
    local defaultConfig = {
        name = "Classic Oval",
        width = 75,
        points = {
            {x = 400, y = 50},
            {x = 700, y = 150},
            {x = 750, y = 300},
            {x = 700, y = 450},
            {x = 400, y = 550},
            {x = 100, y = 450},
            {x = 50, y = 300},
            {x = 100, y = 150},
        },
        startAngle = 0,
        surfaceZones = {
            { startPct = 0.0,  endPct = 0.15, grip = 0.95, bumpiness = 0.05, name = "Smooth Tarmac", color = {0.5, 0.5, 0.5, 0.0} },
            { startPct = 0.15, endPct = 0.30, grip = 0.7,  bumpiness = 0.3,  name = "Worn Patch",    color = {0.6, 0.4, 0.2, 0.08} },
            { startPct = 0.30, endPct = 0.50, grip = 0.95, bumpiness = 0.05, name = "Smooth Tarmac", color = {0.5, 0.5, 0.5, 0.0} },
            { startPct = 0.50, endPct = 0.65, grip = 0.85, bumpiness = 0.6,  name = "Bumpy Section", color = {0.4, 0.35, 0.3, 0.06} },
            { startPct = 0.65, endPct = 0.80, grip = 0.95, bumpiness = 0.05, name = "Smooth Tarmac", color = {0.5, 0.5, 0.5, 0.0} },
            { startPct = 0.80, endPct = 1.0,  grip = 0.6,  bumpiness = 0.1,  name = "Damp Corner",   color = {0.2, 0.3, 0.7, 0.07} },
        },
    }
    track.initFromConfig(defaultConfig)
end

-- Check if a point is on the track (squared distance comparison, no sqrt)
function track.isOnTrack(x, y)
    local minDistSq = huge

    for _, p in ipairs(track.centerPath) do
        local dx = x - p.x
        local dy = y - p.y
        local distSq = dx * dx + dy * dy
        if distSq < minDistSq then
            minDistSq = distSq
        end
    end

    local halfWidth = track.width / 2
    return minDistSq <= halfWidth * halfWidth
end

-- Get the percentage along the track for a given position (squared distance, no sqrt)
function track.getTrackPercent(x, y)
    local minDistSq = huge
    local bestIndex = 1

    for i, p in ipairs(track.centerPath) do
        local dx = x - p.x
        local dy = y - p.y
        local distSq = dx * dx + dy * dy
        if distSq < minDistSq then
            minDistSq = distSq
            bestIndex = i
        end
    end

    return track.cumulative[bestIndex] / track.pathLength
end

function track.getSurfaceAt(x, y)
    local pct = track.getTrackPercent(x, y)
    
    for _, zone in ipairs(track.surfaceZones) do
        if pct >= zone.startPct and pct < zone.endPct then
            return zone
        end
    end
    
    return track.surfaceZones[1]
end

-- Get a point at a specific percentage along the track (0.0 to 1.0)
function track.getPointAtPercent(pct)
    return getPointAtPercent(track.centerPath, track.cumulative, track.pathLength, pct)
end

-- Get approximate track circumference (for stats)
function track.getCircumference()
    return track.pathLength
end

return track

File: crates/sim/src/track.rs
//! Track geometry, surface zones, curbs, and trees (port of track.lua).

use crate::rng::Lcg;
use crate::spline::catmull_rom;
use crate::P;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SurfaceZone {
    #[serde(rename = "startPct")]
    pub start_pct: f64,
    #[serde(rename = "endPct")]
    pub end_pct: f64,
    pub grip: f64,
    pub bumpiness: f64,
    pub name: String,
    /// RGBA overlay tint; alpha 0 = invisible.
    #[serde(default = "default_zone_color")]
    pub color: [f64; 4],
}

fn default_zone_color() -> [f64; 4] {
    [0.5, 0.5, 0.5, 0.0]
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TrackConfig {
    pub id: String,
    pub name: String,
    #[serde(default)]
    pub description: String,
    pub width: f64,
    pub points: Vec<P>,
    #[serde(rename = "startAngle", default)]
    pub start_angle: f64,
    #[serde(rename = "surfaceZones", default)]
    pub surface_zones: Vec<SurfaceZone>,
    #[serde(default)]
    pub seed: i64,
    #[serde(default)]
    pub generated: bool,
}

#[derive(Debug, Clone)]
pub struct Curb {
    pub x: f64,
    pub y: f64,
    pub angle: f64,
    pub index: usize,
}

#[derive(Debug, Clone)]
pub struct Tree {
    pub x: f64,
    pub y: f64,
    pub trunk_h: f64,
    pub canopy_r: f64,
    pub green: f64,
    pub shade: f64,
}

#[derive(Debug, Clone)]
pub struct Track {
    pub config: TrackConfig,
    pub name: String,
    pub width: f64,
    pub center_path: Vec<P>,
    pub path_length: f64,
    pub cumulative: Vec<f64>,
    pub inner_path: Vec<P>,
    pub outer_path: Vec<P>,
    pub finish_x: f64,
    pub finish_y1: f64,
    pub finish_y2: f64,
    pub finish_angle: f64,
    pub finish_point: P,
    /// Track tangent at the finish line = forward direction.
    pub finish_forward: P,
    pub finish_p1: P,
    pub finish_p2: P,
    /// Approximate center of the layout (for trees / legacy behavior).
    pub cx: f64,
    pub cy: f64,
    pub outer_curbs: Vec<Curb>,
    pub inner_curbs: Vec<Curb>,
    pub trees: Vec<Tree>,
    pub surface_zones: Vec<SurfaceZone>,
    pub start_x: f64,
    pub start_y: f64,
    pub start_angle: f64,
}

/// Generate a smooth closed path from control points using Catmull-Rom splines.
pub fn generate_spline_path(control_points: &[P], segments_per_curve: usize) -> Vec<P> {
    let n = control_points.len();
    let mut path = Vec::with_capacity(n * segments_per_curve);
    for i in 0..n {
        let p0 = control_points[(i + n - 1) % n];
        let p1 = control_points[i];
        let p2 = control_points[(i + 1) % n];
        let p3 = control_points[(i + 2) % n];
        for j in 0..segments_per_curve {
            let t = j as f64 / segments_per_curve as f64;
            path.push(catmull_rom(p0, p1, p2, p3, t));
        }
    }
    path
}

fn calculate_path_metrics(path: &[P]) -> (f64, Vec<f64>) {
    let mut total = 0.0;
    let mut cumulative = vec![0.0; path.len()];
    for i in 1..path.len() {
        let dx = path[i].x - path[i - 1].x;
        let dy = path[i].y - path[i - 1].y;
        total += (dx * dx + dy * dy).sqrt();
        cumulative[i] = total;
    }
    // Close the loop
    let dx = path[0].x - path[path.len() - 1].x;
    let dy = path[0].y - path[path.len() - 1].y;
    total += (dx * dx + dy * dy).sqrt();
    (total, cumulative)
}

fn point_at_percent(path: &[P], cumulative: &[f64], total_length: f64, pct: f64) -> P {
    let target = pct * total_length;
    for i in 1..path.len() {
        if cumulative[i] >= target {
            let prev = cumulative[i - 1];
            let seg = cumulative[i] - prev;
            let t = if seg > 0.0 {
                (target - prev) / seg
            } else {
                0.0
            };
            return P {
                x: path[i - 1].x + t * (path[i].x - path[i - 1].x),
                y: path[i - 1].y + t * (path[i].y - path[i - 1].y),
            };
        }
    }
    path[path.len() - 1]
}

/// Unit tangent at a path index (central difference over neighbors).
pub fn tangent_at(path: &[P], index: usize) -> (f64, f64) {
    let n = path.len();
    let prev = path[(index + n - 1) % n];
    let next = path[(index + 1) % n];
    let dx = next.x - prev.x;
    let dy = next.y - prev.y;
    let len = (dx * dx + dy * dy).sqrt();
    if len > 0.0 {
        (dx / len, dy / len)
    } else {
        (1.0, 0.0)
    }
}

impl Track {
    pub fn from_config(config: &TrackConfig) -> Self {
        let center_path = generate_spline_path(&config.points, 25);
        let (path_length, cumulative) = calculate_path_metrics(&center_path);

        let half_width = config.width / 2.0;
        let mut inner_path = Vec::with_capacity(center_path.len());
        let mut outer_path = Vec::with_capacity(center_path.len());
        for (i, p) in center_path.iter().enumerate() {
            let (tx, ty) = tangent_at(&center_path, i);
            let (nx, ny) = (-ty, tx);
            inner_path.push(P {
                x: p.x + nx * half_width,
                y: p.y + ny * half_width,
            });
            outer_path.push(P {
                x: p.x - nx * half_width,
                y: p.y - ny * half_width,
            });
        }

        // Finish line at the start of the track
        let start_point = center_path[0];
        let (tx, ty) = tangent_at(&center_path, 0);
        let (nx, ny) = (-ty, tx);

        let finish_angle = ty.atan2(tx);
        let finish_p1 = P {
            x: start_point.x + nx * half_width,
            y: start_point.y + ny * half_width,
        };
        let finish_p2 = P {
            x: start_point.x - nx * half_width,
            y: start_point.y - ny * half_width,
        };

        // Approximate center for trees
        let (sum_x, sum_y) = config
            .points
            .iter()
            .fold((0.0, 0.0), |(sx, sy), p| (sx + p.x, sy + p.y));
        let cx = sum_x / config.points.len() as f64;
        let cy = sum_y / config.points.len() as f64;

        let surface_zones = if config.surface_zones.is_empty() {
            vec![SurfaceZone {
                start_pct: 0.0,
                end_pct: 1.0,
                grip: 0.95,
                bumpiness: 0.05,
                name: "Smooth Tarmac".into(),
                color: default_zone_color(),
            }]
        } else {
            config.surface_zones.clone()
        };

        let mut track = Track {
            config: config.clone(),
            name: config.name.clone(),
            width: config.width,
            center_path,
            path_length,
            cumulative,
            inner_path,
            outer_path,
            finish_x: start_point.x,
            finish_y1: finish_p1.y,
            finish_y2: finish_p2.y,
            finish_angle,
            finish_point: start_point,
            finish_forward: P { x: tx, y: ty },
            finish_p1,
            finish_p2,
            cx,
            cy,
            outer_curbs: Vec::new(),
            inner_curbs: Vec::new(),
            trees: Vec::new(),
            surface_zones,
            start_x: start_point.x,
            start_y: start_point.y,
            start_angle: if config.start_angle != 0.0 {
                config.start_angle
            } else {
                finish_angle
            },
        };
        track.generate_curbs();
        track.generate_trees();
        track
    }

    /// Legacy oval track (port of track.init()).
    pub fn default_oval() -> Self {
        let mk_zone =
            |start: f64, end: f64, grip: f64, bump: f64, name: &str, color: [f64; 4]| SurfaceZone {
                start_pct: start,
                end_pct: end,
                grip,
                bumpiness: bump,
                name: name.into(),
                color,
            };
        let config = TrackConfig {
            id: "default".into(),
            name: "Classic Oval".into(),
            description: "The classic".into(),
            width: 75.0,
            points: vec![
                P::new(400.0, 50.0),
                P::new(700.0, 150.0),
                P::new(750.0, 300.0),
                P::new(700.0, 450.0),
                P::new(400.0, 550.0),
                P::new(100.0, 450.0),
                P::new(50.0, 300.0),
                P::new(100.0, 150.0),
            ],
            start_angle: 0.0,
            surface_zones: vec![
                mk_zone(0.0, 0.15, 0.95, 0.05, "Smooth Tarmac", [0.5, 0.5, 0.5, 0.0]),
                mk_zone(0.15, 0.30, 0.7, 0.3, "Worn Patch", [0.6, 0.4, 0.2, 0.08]),
                mk_zone(
                    0.30,
                    0.50,
                    0.95,
                    0.05,
                    "Smooth Tarmac",
                    [0.5, 0.5, 0.5, 0.0],
                ),
                mk_zone(
                    0.50,
                    0.65,
                    0.85,
                    0.6,
                    "Bumpy Section",
                    [0.4, 0.35, 0.3, 0.06],
                ),
                mk_zone(
                    0.65,
                    0.80,
                    0.95,
                    0.05,
                    "Smooth Tarmac",
                    [0.5, 0.5, 0.5, 0.0],
                ),
                mk_zone(0.80, 1.0, 0.6, 0.1, "Damp Corner", [0.2, 0.3, 0.7, 0.07]),
            ],
            seed: 0,
            generated: false,
        };
        Self::from_config(&config)
    }

    fn generate_curbs(&mut self) {
        self.outer_curbs.clear();
        self.inner_curbs.clear();
        let step = (self.center_path.len() / 80).max(1);
        let mut index = 0;
        let mut i = 0;
        while i < self.center_path.len() {
            let outer = self.outer_path[i];
            let inner = self.inner_path[i];
            let (tx, ty) = tangent_at(&self.center_path, i);
            let angle = ty.atan2(tx);
            self.outer_curbs.push(Curb {
                x: outer.x,
                y: outer.y,
                angle,
                index,
            });
            self.inner_curbs.push(Curb {
                x: inner.x,
                y: inner.y,
                angle,
                index,
            });
            index += 1;
            i += step;
        }
    }

    fn generate_trees(&mut self) {
        self.trees.clear();
        let mut rng = Lcg::new(77);
        for _ in 0..25 {
            for _attempt in 0..50 {
                let x = rng.int(20, 780) as f64;
                let y = rng.int(20, 580) as f64;
                let mut min_dist = f64::INFINITY;
                for p in &self.center_path {
                    let dx = x - p.x;
                    let dy = y - p.y;
                    min_dist = min_dist.min((dx * dx + dy * dy).sqrt());
                }
                if min_dist > self.width / 2.0 + 15.0 {
                    self.trees.push(Tree {
                        x,
                        y,
                        trunk_h: 6.0 + rng.next() * 4.0,
                        canopy_r: 8.0 + rng.next() * 7.0,
                        green: 0.3 + rng.next() * 0.3,
                        shade: 0.1 + rng.next() * 0.1,
                    });
                    break;
                }
            }
        }
    }

    /// Whether a point is on the track surface.
    pub fn is_on_track(&self, x: f64, y: f64) -> bool {
        let mut min_dist_sq = f64::INFINITY;
        for p in &self.center_path {
            let dx = x - p.x;
            let dy = y - p.y;
            min_dist_sq = min_dist_sq.min(dx * dx + dy * dy);
        }
        let half = self.width / 2.0;
        min_dist_sq <= half * half
    }

    /// Percentage along the track for a given position.
    pub fn get_track_percent(&self, x: f64, y: f64) -> f64 {
        let mut min_dist_sq = f64::INFINITY;
        let mut best = 0;
        for (i, p) in self.center_path.iter().enumerate() {
            let dx = x - p.x;
            let dy = y - p.y;
            let d = dx * dx + dy * dy;
            if d < min_dist_sq {
                min_dist_sq = d;
                best = i;
            }
        }
        self.cumulative[best] / self.path_length
    }

    pub fn get_surface_at(&self, x: f64, y: f64) -> &SurfaceZone {
        let pct = self.get_track_percent(x, y);
        self.surface_zones
            .iter()
            .find(|z| pct >= z.start_pct && pct < z.end_pct)
            .unwrap_or(&self.surface_zones[0])
    }

    pub fn get_point_at_percent(&self, pct: f64) -> P {
        point_at_percent(&self.center_path, &self.cumulative, self.path_length, pct)
    }

    pub fn circumference(&self) -> f64 {
        self.path_length
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tracks::TrackList;

    #[test]
    fn is_on_track_at_start_position() {
        let t = Track::default_oval();
        assert!(t.is_on_track(t.start_x, t.start_y));
    }

    #[test]
    fn infield_is_off_track() {
        let t = Track::default_oval();
        assert!(!t.is_on_track(t.cx, t.cy));
    }

    #[test]
    fn origin_is_off_track() {
        let t = Track::default_oval();
        assert!(!t.is_on_track(0.0, 0.0));
    }

    #[test]
    fn point_along_center_path_is_on_track() {
        let t = Track::default_oval();
        let p = t.center_path[9];
        assert!(t.is_on_track(p.x, p.y));
    }

    #[test]
    fn surface_zone_has_valid_properties() {
        let t = Track::default_oval();
        let zone = t.get_surface_at(t.start_x, t.start_y);
        assert!(zone.grip > 0.0);
        assert!(!zone.name.is_empty());
    }

    #[test]
    fn generates_curbs_and_trees() {
        let t = Track::default_oval();
        assert!(!t.outer_curbs.is_empty());
        assert!(!t.inner_curbs.is_empty());
        assert!(!t.trees.is_empty());
    }

    #[test]
    fn surface_zones_cover_zero_to_one() {
        let t = Track::default_oval();
        assert!(t.surface_zones[0].start_pct.abs() < 1e-3);
        assert!((t.surface_zones.last().unwrap().end_pct - 1.0).abs() < 1e-3);
    }

    #[test]
    fn inner_and_outer_paths_match_center_length() {
        let list = TrackList::with_defaults();
        let t = Track::from_config(list.get_by_index(0).unwrap());
        assert_eq!(t.inner_path.len(), t.center_path.len());
        assert_eq!(t.outer_path.len(), t.center_path.len());
        assert!(!t.center_path.is_empty());
    }

    #[test]
    fn track_percent_is_in_unit_range() {
        let list = TrackList::with_defaults();
        let t = Track::from_config(list.get_by_index(0).unwrap());
        let pct = t.get_track_percent(t.start_x, t.start_y);
        assert!((0.0..=1.0).contains(&pct));
    }

    #[test]
    fn circumference_is_positive() {
        let list = TrackList::with_defaults();
        let t = Track::from_config(list.get_by_index(0).unwrap());
        assert!(t.circumference() > 0.0);
    }

    #[test]
    fn all_default_tracks_initialize_with_start_on_track() {
        let list = TrackList::with_defaults();
        for cfg in &list.list {
            let t = Track::from_config(cfg);
            assert!(!t.center_path.is_empty(), "{} has no path", cfg.name);
            assert!(
                t.is_on_track(t.start_x, t.start_y),
                "{} start pos not on track",
                cfg.name
            );
        }
    }

    #[test]
    fn get_point_at_percent_walks_the_loop() {
        let t = Track::default_oval();
        let p0 = t.get_point_at_percent(0.0);
        let p50 = t.get_point_at_percent(0.5);
        let dx = p50.x - p0.x;
        let dy = p50.y - p0.y;
        assert!(
            (dx * dx + dy * dy).sqrt() > t.width,
            "halfway point should be far from start"
        );
    }
}
